### PR TITLE
[New Ivy Function] `sliding_window`

### PR DESCRIPTION
**Pull request by adityagandhamal**
Working on https://trello.com/c/C6ZfGJh3/2209-implement-ivyslidingwindow-to-optimize-torchunfold-and-ivyreducewindow

### CODE REVIEW

File: ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_layers.py [hunk 1]
```diff
@@ -1276,3 +1322,25 @@ def test_rfftn(
...
         norm=norm,
     )
+
+
+@handle_test(
+    fn_tree="functional.ivy.experimental.sliding_window",
+    all_args=_sliding_window_helper(),
+    test_with_out=st.just(False),
+    ground_truth_backend="jax",
+)
```
Comment (AnnaTz) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_layers.py:1332]: In the CI checks I see some failures that I think are related to the missing array and container methods. You need to either add the needed code, or exclude these cases from `handle_test` and add a ToDo.

Other than that, looks great overall!

**Review by AnnaTz — CHANGES_REQUESTED** [review]

File: ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_layers.py [hunk 2]
```diff
@@ -1276,3 +1322,25 @@ def test_rfftn(
...
+def test_sliding_window(*, all_args, test_flags, backend_fw, fn_name, on_device):
```
Comment (AnnaTz) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_layers.py:1333]: Sadly I can still see some strange errors in the Ci checks. It might be that they didn't get updated for some reason though... Have you tried running the test locally?

File: ivy/data_classes/array/experimental/layers.py [hunk 3]
```diff
@@ -1117,3 +1117,57 @@ def rfftn(
...
+    def sliding_window(
+        self: ivy.Array,
+        window_size: Union[int, Tuple[int, int], Tuple[int, int, int]],
+        stride: Union[int, Tuple[int, int]] = 1,
+        dilation: Union[int, Tuple[int, int]] = 1,
+        padding: Union[str, int, Sequence[Tuple[int, int]]] = "VALID",
+        /,
+        *,
+        data_format="NCHW",
+    ) -> ivy.Array:
```
Comment (AnnaTz) [inline ivy/data_classes/array/experimental/layers.py:1130]: Also, since `stride`, `dilation` and `padding` have default values too, I think the `/, *` should be before them.

**Review by AnnaTz — CHANGES_REQUESTED** [review]

File: ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_layers.py (hunk 2 above)
Comment (adityagandhamal) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_layers.py:1333]: Yeah, the tests are failing on my side as well

File: ivy/data_classes/array/experimental/layers.py (hunk 3 above)
Comment (adityagandhamal) [reply ivy/data_classes/array/experimental/layers.py:1130]: yes, I'll make the changes accordingly

**Comment by adityagandhamal** [comment]
Hey @AnnaTz , looking at the logs, I think we need to look at the args generated by the test cases here. I have referred to the test function of the original `reduce_window` method for this one and it too fails in certain cases.

I mean I've tried the demos on every backend, seem to run fine and so is the case with the optimized `reduce_window` and `torch_frontend.unfold` methods.

### Backend Numpy

```
import ivy
ivy.set_backend('numpy')

input = ivy.array([[1, 2, 3],
                [4, 5, 6],
                [7, 8, 9]])

print(ivy.sliding_window(input, (2, 2)))

ivy.array([[[1, 2, 4, 5],
        [2, 3, 5, 6]],

       [[4, 5, 7, 8],
        [5, 6, 8, 9]]])
```
### Backend Jax

```
import ivy
import jax
jax.config.update('jax_enable_x64', True)

ivy.set_backend('jax')

input = ivy.array([[1, 2, 3],
                [4, 5, 6],
                [7, 8, 9]])

print(ivy.sliding_window(input, (2, 2)))

ivy.array([[[1, 2, 4, 5],
        [2, 3, 5, 6]],

       [[4, 5, 7, 8],
        [5, 6, 8, 9]]])
```

### Backend Torch
```
import ivy

ivy.set_backend('torch')

input = ivy.array([[1, 2, 3],
                [4, 5, 6],
                [7, 8, 9]])

print(ivy.sliding_window(input, (2, 2)))

ivy.array([[[1., 2., 4., 5.],
        [2., 3., 5., 6.],
        [4., 5., 7., 8.],
        [5., 6., 8., 9.]]])
```
### Backend Tensorflow
```
import ivy

ivy.set_backend('tensorflow')

input = ivy.array([[1, 2, 3],
                [4, 5, 6],
                [7, 8, 9]])


print(ivy.sliding_window(input, (2, 2)))

ivy.array([[[[1, 2, 4, 5],
         [2, 3, 5, 6]],

        [[4, 5, 7, 8],
         [5, 6, 8, 9]]]])
```

### Reduce Window Op

```
import ivy

ivy.set_backend('tensorflow')

input = ivy.array([[1, 2, 3],
                [4, 5, 6],
                [7, 8, 9]])


print(ivy.reduce_window(input, 0, ivy.add, (2, 2)))

ivy.array([[[12, 16],
        [24, 28]]])
```

### Torch Frontend Unfold
```
import ivy
import ivy.functional.frontends.torch as torch_frontend
ivy.set_backend('tensorflow')

input = ivy.array([[[[1, 2, 3],
                [4, 5, 6],
                [7, 8, 9]]]])


print(torch_frontend.nn.functional.unfold(input, (2, 2)))

ivy.frontends.torch.Tensor([[[[1, 2, 4, 5],
         [2, 3, 5, 6]],

        [[4, 5, 7, 8],
         [5, 6, 8, 9]]]])
```

File: ivy/functional/backends/torch/experimental/layers.py [hunk 4]
```diff
@@ -1010,3 +1010,45 @@ def rfftn(
...
+        # convert padding from str to seq
+        if padding.upper() == "SAME":
+            pad_vals = []
+            for dim in input.shape:
+                pad_val = _handle_padding(
+                    dim,
+                    stride[0] if isinstance(stride, tuple) else stride,
+                    kernel_size[0],
+                    padding,
+                )
```
Comment (AnnaTz) [inline ivy/functional/backends/torch/experimental/layers.py:1042]: Hi @adityagandhamal, I see this failure in the CI checks:
image

File: ivy/functional/backends/numpy/experimental/layers.py [hunk 5]
```diff
@@ -1044,3 +1049,33 @@ def rfftn(
...
+
+    input = input.reshape((1, 1) + input.shape)
+    if dilation:
+        identity = ivy.array(0)
+        input = _dilate(input, dilation, identity)
+
+    view = _conv_view(input, [1, 1] + list(k_size), stride, pads, identity)[0]
+    view = ivy.reshape(view, (*view.shape[1 : 1 + len(k_size)], -1))
+
+    return view
```
Comment (AnnaTz) [inline ivy/functional/backends/numpy/experimental/layers.py:1081]: Currently you haven't added a paddle backend at all. You could move this compositional implementation to ivy and thus convert the function to a mixed function.
Also, since the jax backend is pretty much the same, we can omit that as well and have the function fall back to the main ivy composition. This will avoid code repetition.
So basically, we would end up with the torch and tensorflow backends, plus an ivy implementation that will run for the rest.

File: ivy/functional/backends/tensorflow/experimental/layers.py [hunk 6]
```diff
@@ -1427,3 +1427,54 @@ def rfftn(
...
+        if padding[0] == padding[1] == 0:
+            padding = "VALID"
+        elif padding[0] == padding[1] != 0:
+            padding = "SAME"
+        else:
+            raise ivy.utils.exceptions.IvyError(
+                f"Cannot convert padding sequence {padding} to TensorFlow padding mode"
+            )
+
+    return tf.image.extract_patches(input, kernel_size, stride, dilation, padding)
```
Comment (AnnaTz) [inline ivy/functional/backends/tensorflow/experimental/layers.py:1480]: The other error I'm seeing in the CI checks seems to be specific to tensorflow:
image
You'll have to debug this locally. I don't think this is caused by the test because it seems it only occurs in this backend.

**Review by AnnaTz — CHANGES_REQUESTED** [review]

File: ivy/functional/backends/tensorflow/experimental/layers.py (hunk 6 above)
Comment (adityagandhamal) [reply ivy/functional/backends/tensorflow/experimental/layers.py]: This bit needed a slight change in the way tensorflow accepts `kernel_size`. This is resolved with the latest commits.

File: ivy/functional/backends/numpy/experimental/layers.py (hunk 5 above)
Comment (adityagandhamal) [reply ivy/functional/backends/numpy/experimental/layers.py]: Have found out a native function for paddle backend in https://www.paddlepaddle.org.cn/documentation/docs/en/api/paddle/nn/functional/unfold_en.html. Also, the latest commits include the changes as suggested.

File: ivy/functional/backends/torch/experimental/layers.py (hunk 4 above)
Comment (adityagandhamal) [reply ivy/functional/backends/torch/experimental/layers.py:1049]: isn't the case anymore with the latest commits

**Comment by adityagandhamal** [comment]
Hey @AnnaTz, I've made the changes to the earlier implementation by making `ivy.sliding_window` a mixed function with the backends torch, tensorflow, and paddle and an ivy implementation for numpy and jax.

**Comment by adityagandhamal** [comment]
Hey @AnnaTz, I've made the changes to the earlier implementation by making `ivy.sliding_window` a mixed function with the backends torch, tensorflow, and paddle and an ivy implementation for numpy and jax.

**Comment by AnnaTz** [comment]
Hey @adityagandhamal, I think you have removed/replaced `ivy.stft` with your code. Could you look into that?

**Review by AnnaTz — CHANGES_REQUESTED** [review]
Everything looks good now, except for the `stft` issue. Was there a reason to remove it that I am missing?

**Comment by adityagandhamal** [comment]
> Everything looks good now, except for the `stft` issue. Was there a reason to remove it that I am missing?

My bad. This might have been caused while resolving the merge conflicts before. Pulling from `main` should resolve the issue

**Comment by AnnaTz** [comment]
> > Everything looks good now, except for the `stft` issue. Was there a reason to remove it that I am missing?
> 
> My bad. This might have been caused while resolving the merge conflicts before. Pulling from `main` should resolve the issue

Seems it didn't resolve it. I still see `stft` being erased in the `Files changed` tab.

**Comment by adityagandhamal** [comment]
> Seems it didn't resolve it. I still see `stft` being erased in the `Files changed` tab. Checked your branch as well. `stft` is missing there.

Not sure why is this the case though but anyways I copied all the elements pertaining to `stft` from `main`

File: ivy/functional/backends/jax/experimental/layers.py [hunk 7]
```diff
@@ -235,9 +237,9 @@ def max_pool2d(
     )
 
     if data_format == "NCHW":
-        res = jnp.transpose(res, (0, 3, 1, 2))
+        return jnp.transpose(res, (0, 3, 1, 2))
 
-    return res.astype(odtype)
+    return res
```
Comment (AnnaTz) [inline ivy/functional/backends/jax/experimental/layers.py:243]: Okay last thing, are these changes to `max_pool2d` related to the PR? If not you should make a separate commit.

**Review by AnnaTz — CHANGES_REQUESTED** [review]

File: ivy/functional/backends/jax/experimental/layers.py (hunk 7 above)
Comment (adityagandhamal) [reply ivy/functional/backends/jax/experimental/layers.py:243]: No, will revert this to the original logic

File: ivy/functional/backends/jax/experimental/layers.py [hunk 8]
```diff
@@ -235,7 +238,7 @@ def max_pool2d(
     )
 
     if data_format == "NCHW":
-        res = jnp.transpose(res, (0, 3, 1, 2))
+        return jnp.transpose(res, (0, 3, 1, 2))
```
Comment (AnnaTz) [inline ivy/functional/backends/jax/experimental/layers.py:241]: You didn't completely revert the changes 😅

**Review by AnnaTz — APPROVED** [review]
Thanks!

File: ivy/functional/backends/jax/experimental/layers.py (hunk 8 above)
Comment (adityagandhamal) [reply ivy/functional/backends/jax/experimental/layers.py]: :neutral_face:
